File: uvm/hier/usr/share/untangle/bin/ut-createUID.rb
#! /usr/bin/ruby1.8

require 'gpgme'
require 'optparse'

# The format of a UID is:
#   xxxx-xxxx-xxxx-xxPV-xxxxxxx-xxxxxxx-xxxxPVx
#   P = platform bit
#   V = version bit
#   x = uuid bit
# see VERSIONS and PLATFORMS below for the actual meaning of those
# bits

#############################
# constants

# bits
PLATFORMS = { :sarge    => 0,
              :etch     => 1,
              :sid      => 2,
              :feisty   => 3,
              :gutsy    => 4,
              :hardy    => 5,
              :intrepid => 6,
              :lenny    => 7,
              :squeeze  => 8 }

VERSIONS = { :hardware => 1,
             :iso      => 2,
             :pkgs     => 3,
             :win      => 4,
             :mac      => 5,
             :lin      => 6,
             :ubuntu   => 9 }

# workaround for <1.0 gpgme
OLDGPGME = !defined?(GPGME::Ctx)

# files
UNTANGLE = "@PREFIX@/usr/share/untangle"
UNTANGLE_GPG_HOME = File.join(UNTANGLE, "gpg")
UID_FILE = File.join(UNTANGLE, "conf/uid")
ENTROPY_AVAILABLE_FILE = "/proc/sys/kernel/random/entropy_avail"

# minimum entropy needed to generate our key
MINIMUM_ENTROPY = 3000

# length of the left part
LEFT_LENGTH = 16

# default GPG key config
GPG_KEY_CONFIG = <<EOF
<GnupgKeyParms format="internal">
Key-Type: RSA
Key-Length: 2048
Name-Real: Untangle
Name-Email: untangle@untangle.com
Expire-Date: 0
</GnupgKeyParms>
EOF

#############################
# functions

def parseCommandLineArgs(args)
  options = { :typeNibble => nil }
  
  opts = OptionParser.new { |opts|
    opts.on("--type-nibble=TYPE", VERSIONS.keys, "force the type nibble #{VERSIONS.keys.map {|k| k.to_s}.inspect}") { |t|
      options[:typeNibble] = VERSIONS[t]
    }
  }

  begin
    opts.parse!(args)
  rescue Exception => e
    puts e, "", opts
    exit
  end
  
  return options
end

def hasEnoughEntropy
  return (File.new(ENTROPY_AVAILABLE_FILE).read.to_i > MINIMUM_ENTROPY)
end

def createGpgKeyring(config)
  if hasEnoughEntropy && !OLDGPGME then 
    puts "GPGME"
    createGpgKeyringWithGPGME(config)
  else
    puts "GPG"
    createGpgKeyringWithGPG(config)
  end
end

def createGpgKeyringWithGPGME(config)
  ctx = GPGME::Ctx.new
  ctx.genkey(config, nil, nil)
end

def createGpgKeyringWithGPG(config)
  IO.popen("gpg --batch --quick-random --gen-key 2> /dev/null", 'w+') { |f|
    f.puts(config.gsub!(/^<.*/, '')) # gpg doesn't want the XML-style config
  }
  # exceptions occur for lots of reasons, like running the first time.
  # raise Exception.new if $? != 0
end

def getPlatform
  case File.open("/etc/issue").read
  when %r|lenny/sid|
    :sid
  when /4.0/
    :etch
  when /3\.1/
    :sarge
  when /5\.0/
    :lenny
  when /6\.0/
    :squeeze
  when /(7\.04|feisty)/
    :feisty
  when /(7\.10|gutsy)/
    :gutsy
  when /(8\.04|hardy)/
    :hardy
  when /(8\.10|intrepid)/
    :intrepid
  else
    raise "Couldn't get your platform."
  end
end

# ACTIVATION_FILE = File.join(UNTANGLE, "activation.key")

# def readExistingActivationKey
#   existingLicenseKey = nil
#   if File.file?(ACTIVATION_FILE) then
#     existingLicenseKey = File.open(ACTIVATION_FILE).read.strip.gsub(/-/, "")
#     existingLicenseKey = nil if existingLicenseKey =~ /^[0\-]+$/
#   end
#   return existingLicenseKey
# end

def getPackageVersion(name)
  line = `dpkg-query --showformat='${Version}' -W #{name} 2> /dev/null`
  version = line.empty? ? nil : line
  return version
end

def getVersion
  if !getPackageVersion('untangle-windows-installer').nil? 
    :win
  elsif !getPackageVersion('untangle-mac-installer').nil? 
    :mac
  elsif !getPackageVersion('untangle-linux-installer').nil? 
    :lin
  elsif !getPackageVersion('untangle').nil? 
    :ubuntu
  elsif getPackageVersion('untangle-gateway').nil?
    :pkgs
  elsif getPackageVersion('untangle-hardware-support').nil?
    :iso
  else
    :hardware
  end
end

def getBits(typeNibble)
  typeNibble = VERSIONS[getVersion] if typeNibble.nil?
  return "#{PLATFORMS[getPlatform]}#{typeNibble}"
end

def createUID(fingerprint, bits, existingKey)
  # insert meaningful bit in the left part
  if existingKey.nil? then
    uid = fingerprint.insert(LEFT_LENGTH-2, bits)
  else # FIXME
    uid = existingKey + fingerprint[LEFT_LENGTH-2,fingerprint.length-LEFT_LENGTH+2]
  end

  # insert meaningful bit in the right part
  uid = uid.insert(-2, bits).downcase

  # group and join
  one = uid[0,LEFT_LENGTH].scan(/.{4}/)
  two = uid[LEFT_LENGTH,uid.length-LEFT_LENGTH].scan(/.{7}/)
  return [one, one+two].map{ |list| list.join("-") }
end

def writeToFiles(shortUID, uid)
#  File.open(ACTIVATION_FILE, 'w') { |f| f.puts(shortUID) }
  File.open(UID_FILE, 'w') { |f| f.puts(shortUID) }
end

#######################
# main

options = parseCommandLineArgs(ARGV)

# GPG env.
ENV['GPG_AGENT_INFO'] = nil
ENV['GNUPGHOME'] = UNTANGLE_GPG_HOME

# create the keyring only if no keys are available
firsttime = true
begin
  fingerprint = nil
  if (OLDGPGME)
    IO.popen("gpg --with-colons --fingerprint 2> /dev/null") { |f|
      f.each do |input|
        if (input.index('fpr') == 0)
          arr = input.split(':');
          if (arr.length > 1)
            fingerprint=arr[arr.length - 2];
            break
          end
        end
      end
    }
    raise NoMethodError if fingerprint == nil
  else
    key = GPGME.list_keys[0].subkeys[0]
    fingerprint = key.fingerprint
  end
rescue NoMethodError
  if (firsttime)
    firsttime = false
    createGpgKeyring(GPG_KEY_CONFIG) # create the keyring
    retry
  end
  puts "unable to get fingerprint"
  exit 1
end

bits = getBits(options[:typeNibble]) # what we'll embed in the uid

# existingLicenseKey = readExistingActivationKey # re-use that if it's there

# shortUID, uid = createUID(fingerprint, bits, existingLicenseKey)
shortUID, uid = createUID(fingerprint, bits, nil)

puts shortUID, uid
writeToFiles(shortUID, uid)

File: uvm/hier/usr/share/untangle/bin/ut-createUID.py
#!/usr/bin/python3

# This file creates a UID 
# and writes the uid file  (default: /usr/share/untangle/conf/uid)

import getopt
import random
import sys
import os
import re
import datetime
import traceback

UID_FILENAME=None
class ArgumentParser(object):
    def __init__(self):
        pass

    def set_uidfile( self, arg ):
        global UID_FILENAME
        UID_FILENAME = arg

    def parse_args( self ):
        handlers = {
            '-f' : self.set_uidfile
        }

        try:
            (optlist, args) = getopt.getopt(sys.argv[1:], 'd:n:f:a:u:')
            for opt in optlist:
                handlers[opt[0]](opt[1])
            return args
        except getopt.GetoptError as exc:
            print(exc)
            printUsage()
            exit(1)

def printUsage():
    sys.stderr.write( """\
%s Usage:
  required args:
    -f <uid_filename>          example: "/usr/share/untangle/conf/uid"
""" % (sys.argv[0]) )
    sys.exit(1)

parser = ArgumentParser()
parser.parse_args()

if UID_FILENAME == None:
    printUsage()

# find debian version
debian_distro='wheezy'
try:
    debian_version_file = open( "/etc/debian_version", "r+" )
    ver = debian_version_file.read()
    if re.match(r'^7\.*',ver) != None:
        debian_distro='wheezy'
    elif re.match(r'^8\.*',ver) != None:
        debian_distro='jessie'
    elif re.match(r'^9\.*',ver) != None:
        debian_distro='stretch'
    elif re.match(r'^10\.*',ver) != None:
        debian_distro='buster'
    elif re.match(r'^11\.*',ver) != None:
        debian_distro='bullseye'
    else:
        print("Unknown debian Version %s. Assuming \"%s\"" % (ver.strip(), debian_distro))
except Exception as e:
    traceback.print_exc()
    print("Unknown debian Version %s. Assuming \"%s\"" % (ver.strip(), debian_distro))

# last two bytes in UID have special meaning 
platforms = {'sarge': '0',
             'etch': '1',
             'sid': '2',
             'jessie': '3',
             'stretch': '4',
             'buster': '5',
             'bullseye': '6',
             'lenny': '7',
             'squeeze': '8',
             'wheezy': '9'}
versions = {'hardware': '1', 'iso': '2'}

# generate UID (mostly random bytes)
uid_quad1 = ''.join(random.choice('0123456789abcdef') for i in range(4))
uid_quad2 = ''.join(random.choice('0123456789abcdef') for i in range(4))
uid_quad3 = ''.join(random.choice('0123456789abcdef') for i in range(4))
uid_quad4 = ''.join(random.choice('0123456789abcdef') for i in range(2)) + platforms[debian_distro] + versions['iso']
uid = uid_quad1 + "-" + uid_quad2 + "-" + uid_quad3 + "-" + uid_quad4

# write UID file
file = open( UID_FILENAME, "w+" )
file.write(uid + "\n")
file.flush()
file.close()

sys.exit(0)
